catalyst/callbacks/metric.py

```diff
@@ -1,66 +1,132 @@
 # @TODO: add metric aggregation, etc callback
+from abc import abstractmethod, ABC
+from typing import Dict, Iterable, Union, Tuple
+
+import torch
+
 from catalyst.core.callback import Callback, CallbackNode, CallbackOrder
 from catalyst.core.runner import IRunner
-from catalyst.metrics._metric import ICallbackBatchMetric, ICallbackLoaderMetric
+from catalyst.metrics import ICallbackBatchMetric, ICallbackLoaderMetric, IMetric
 
 
-class IMetricCallback(Callback):
+class IMetricCallback(Callback, ABC):
     """Metric callback interface, abstraction over metric step."""
+    @abstractmethod
+    def on_loader_start(self, runner: "IRunner") -> None:
+        pass
+
+    @abstractmethod
+    def on_batch_end(self, runner: "IRunner") -> None:
+        pass
 
-    pass
+    @abstractmethod
+    def on_loader_end(self, runner: "IRunner") -> None:
+        pass
 
 
-# @TODO: add KV support for input/output
-class BatchMetricCallback(IMetricCallback):
+class MetricCallback(IMetricCallback):
     def __init__(
-        self,
-        metric: ICallbackBatchMetric,
-        input_key: str,
-        target_key: str,
-        log_on_batch: bool = True,
+            self,
+            metric: Union[ICallbackBatchMetric, ICallbackLoaderMetric],
+            input_key: Union[str, Iterable[str], Dict[str, str]],
+            target_key: str,
     ):
         super().__init__(order=CallbackOrder.metric, node=CallbackNode.all)
-        assert isinstance(metric, ICallbackBatchMetric)
         self.metric = metric
+        self._validate_metric(metric=metric)
         self.input_key = input_key
         self.target_key = target_key
+        self._is_value_input = isinstance(input_key, str)
+        self._keys = {
+            **self._format_keys_to_kv(input_key),
+            **self._format_keys_to_kv(target_key),
+        }
+
+    @staticmethod
+    def _validate_metric(metric: IMetric) -> None:
+        assert isinstance(metric, IMetric)
+
+    @staticmethod
+    def _format_keys_to_kv(keys: Union[str, Iterable[str], Dict[str, str]]) -> Dict[str, str]:
+        kv_keys = {}
+        if isinstance(keys, dict):
```

Comment:
could u please check what happens in the `OrderedDict` case?

Reply:
isinstance(OrderedDict(), dict) is True

Reply:
🤔  what about python 3.6?)